consensus/polybft/state_event_getter.go

```diff
@@ -1,12 +1,154 @@
 package polybft
 
 import (
+	"sync"
+
 	"github.com/0xPolygon/polygon-edge/blockchain"
 	"github.com/0xPolygon/polygon-edge/consensus/polybft/contractsapi"
 	"github.com/0xPolygon/polygon-edge/types"
 	"github.com/umbracle/ethgo"
 )
 
+// EventSubscriber specifies functions needed for a component to subscribe to eventProvider
+type EventSubscriber interface {
+	GetLogFilters() map[types.Address][]types.Hash
+	AddLog(header *types.Header, log *ethgo.Log, dbTx DBTransaction) error
+}
+
+// EventProvider represents an event provider in a blockchain system
+// that returns desired events to subscribers from finalized blocks
+// Note that this provider needs to be called manually on each block finalization.
+// The EventProvider struct has the following fields:
+// - blockchain: A blockchainBackend type that represents the blockchain backend used by the event provider.
+// - subscribers: A map[string]eventSubscriber that stores the subscribers of the event provider.
+// - allFilters: A map[types.Address]map[types.Hash][]string that stores the filters for event logs.
+type EventProvider struct {
+	lock                sync.Mutex
+	subscriberIDCounter uint64
+
+	blockchain blockchainBackend
+
+	subscribers map[uint64]EventSubscriber
+	allFilters  map[types.Address]map[types.Hash][]uint64
+}
+
+// NewEventProvider returns a new instance of eventProvider
+func NewEventProvider(blockchain blockchainBackend) *EventProvider {
+	return &EventProvider{
+		blockchain:  blockchain,
+		subscribers: make(map[uint64]EventSubscriber, 0),
+		allFilters:  make(map[types.Address]map[types.Hash][]uint64, 0),
+	}
+}
+
+// Subscribe subscribes given EventSubscriber to desired logs (events)
+func (e *EventProvider) Subscribe(subscriber EventSubscriber) {
+	e.lock.Lock()
+	e.subscriberIDCounter++
+	subscriberID := e.subscriberIDCounter
+	e.subscribers[e.subscriberIDCounter] = subscriber
+	e.lock.Unlock()
```

Comment:
Why do you need lock here? From what I can see, `Subscribe` is called from a single goroutine.

Also use `subscriberID` variable since you have introduced it:

```suggestion
	e.subscriberIDCounter++
	subscriberID := e.subscriberIDCounter
	e.subscribers[subscriberID] = subscriber
```

Reply:
Fix: https://github.com/0xPolygon/polygon-edge/pull/1972/commits/12cd8d0769aeb15c9635ee45aed989c59e3aa7ff